Make modprobe module check for builtins as well

cognifloyd opened the pull request:
Without this modprobe always reports changed when modprobe-ing a builtin module.

##### SUMMARY
This makes the modprobe module check `/usr/lib/$kernel_release/modules.builtin` to avoid reporting changed when state is `present` and the module is builtin.

When the module is builtin, it is already loaded and running modprobe can't change anything (neither adding or removing the kernel module). So, if it's builtin, this should never report changed.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
modprobe module

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
devel
```

cognifloyd committed:
Make modprobe module check for builtins as well

Without this modprobe always reports changed when modprobe-ing a builtin module.

With this, if a kernel module is a builtin, the modprobe module will:
- succeed (without incorrectly reporting changed) if ``state`` is ``present``;
- fail if ``state`` is ``absent``

The failure will have whatever error message modprobe returns when
attempting to remove a builtin module. For example:
``modprobe: ERROR: Module nfs is builtin.``

cognifloyd commented:
from irc:

(16:16:13) alikins: cognifloyd: sounds bugfixy to me.  What should happen if you 'state=absent' a builtin module?
(16:18:23) cognifloyd: It will fail.
(16:18:34) cognifloyd: rmmod will report a helpful message saying it is builtin.
```
⚡ root@pahoran $ modprobe -r nfs
modprobe: ERROR: Module nfs is builtin.
```
(16:19:45) cognifloyd: rc=1

https://github.com/ansible/ansible/blob/eb4db068aa5aa50e4ee2a94b1a036e38a84f1a4f/lib/ansible/modules/system/modprobe.py#L115-L120

Then it's up to the user how to handle a builtin module error appropriately if they are trying to use state=absent.

abadger commented:
Looks like this needs a rebase but the idea here seems sound to me.  There's a whiff of kludge about having to massage the data so much but I can't think of a better way to do it.

Since this has the potential to break existing playbooks we better have a changelog ( https://github.com/ansible/ansible/tree/devel/changelogs/fragments )  and porting guide entry: https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/porting_guides/porting_guide_2.7.rst#noteworthy-module-changes  In the porting guide, be sure to mention that builtins can now make the module fail and how to work around that.

cognifloyd commented:
Rebased. Added changelog entry. And added an entry in the 2.8 porting guide. Will this get backported to 2.7 as well? (ie should it be in the 2.7 porting guide?)

acozine commented:
@cognifloyd since it's a change to functionality, we will not backport to the `stable-2.7` branch. There's another merge conflict from the porting guide - I'm going to try rebasing this; if everything goes smoothly with the rebase and testing, I will merge the PR today.

winem commented:
This does not work for me when I run ansible from the latest devel on Ubuntu 16.04. I experienced this error on target hosts running Ubuntu 16.04 and 18.04. 

Issue is that there is a trailing newline for `uname_kernel_release` on line 95. 

Error message: 

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: FileNotFoundError: [Errno 2] No such file or directory: '/lib/modules/4.4.0-141-generic\n/modules.builtin'
failed: [do-ffm-test-1604] (item=sctp) => {
    "changed": false, 
    "item": "sctp", 
    "name": "sctp", 
    "params": "", 
    "state": "present"
}

MSG:

[Errno 2] No such file or directory: '/lib/modules/4.4.0-141-generic\n/modules.builtin'
```

Task: 

```
- name: Unload modules on the blacklist
  become: true 
  modprobe: 
    name: "{{ item }}"
    state: present
  loop: "{{ modprobe_blacklist }}"
  ignore_errors: true
  tags: 
    - modprobe

# vars/blacklist_modules.yml
---
modprobe_blacklist: 
  - dccp 
  - hfs
  - hfsplus 
  - rds
  - sctp 
  - tipc
```
